include/storage/flash_map.h

```diff
@@ -262,15 +267,63 @@ uint8_t flash_area_erased_val(const struct flash_area *fa);
 #define FLASH_AREA_LABEL_STR(lbl) \
 	DT_PROP(DT_NODE_BY_FIXED_PARTITION_LABEL(lbl), label)
 
-#define FLASH_AREA_ID(label) \
-	DT_FIXED_PARTITION_ID(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
+/**
+ * Obtain pointer to flash_area objet directly from DTS label of fixed partition.
+ *
+ * @param[in] label Fixed partition label to use; this C identifier type label, which menas that
+ *	      instead of for example image-0, image_0 sh uld be provided.
+
+ * @return const struct flash_area * to flash_area object or NULL.
+ */
+#define FLASH_AREA(label)								\
+	(&_CONCAT(_flash_map_area_,							\
+		   DT_FIXED_PARTITION_ID(DT_NODE_BY_FIXED_PARTITION_LABEL(label))))
+
+/**
+ * Obtain identifier of flash_area from DTS label.
+ *
+ * @param[in] label Fixed partition label to use; this C identifier type label, which menas that
+ *	      instead of for example image-0, image_0 sh uld be provided.
+ *
+ * @return Identifier of fixed partition that has been assigned to the partition with given label.
+ */
+#define FLASH_AREA_ID(label) DT_FIXED_PARTITION_ID(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
 
 #define FLASH_AREA_OFFSET(label) \
 	DT_REG_ADDR(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
 
 #define FLASH_AREA_SIZE(label) \
 	DT_REG_SIZE(DT_NODE_BY_FIXED_PARTITION_LABEL(label))
 
+/**
+ * Define flash_area type object and place within _flash_map linker section.
+ *
+ * Usage: DEFINE_FLASH_AREA(some) = { ... };
+ */
+#define DEFINE_FLASH_AREA(name)									\
+	Z_DECL_ALIGN(const struct flash_area) _CONCAT(_flash_map_area_, name)			\
+		__in_section(_flash_map, static, _CONCAT(_flash_map_area_, name)) __used;	\
+	const struct flash_area _CONCAT(_flash_map_area_, name)
+
+/**
+ * Declares flash_area type object for access within code.
+ *
+ * Usage: DECLARE_FLASH_AREA(some);
+ */
+#define DECLARE_FLASH_AREA(name) const extern struct flash_area _CONCAT(_flash_map_area_, name)
+
+
+#define DT_FLASH_AREA_FOREACH_OKAY_FIXED_PARTITION(fn)				\
+		    UTIL_CAT(DT_FOREACH_OKAY_INST_, fixed_partitions)(fn)
```

Comment:
This is an abstraction violation of devicetree.h implementation details -- we shouldn't be using `DT_FOREACH_OKAY_INST_fixed_partitions` like this outside of devicetree.h or its sub-headers.

I see these uses:

 - in flash_map_default.c
- later on in flash_map.h, to declare partitions generated from DTS

The flash_map_default.c case already has defined `DT_DRV_COMPAT` to `fixed_partitions`, so it can simply use DT_INST_FOREACH_STATUS_OKAY without needing this new macro.

If this is really needed here (why?), then I we should instead add a new DT API macro, like this:

```
#define DT_FOREACH_STATUS_OKAY(compat, fn)			\
	COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(compat),		\
		    (UTIL_CAT(DT_FOREACH_OKAY_INST_,		\
			      compat)(fn)),			\
		    ())
```

And then redefine

```
#define DT_INST_FOREACH_STATUS_OKAY(fn)				\
	DT_FOREACH_STATUS_OKAY(DT_DRV_COMPAT, fn)
```

to avoid the abstraction violation.

Reply:
I wanted to avoid adding `DT_DRV_COMPAT ` to header because a source that includes the `flash_map.h` maybe defining the name before or after inclusion of flash_map which may give error during compilation and if I undef it here, to define my own then the rest of a compilation unit will have either no definition when needed or incorrect one.

Maybe the `DT_INST_FOREACH_STATUS_OKAY` is required.

Reply:
> Maybe the `DT_INST_FOREACH_STATUS_OKAY` is required.

You meant `DT_FOREACH_STATUS_OKAY`, right?